File: handlers/hostinfo/disk.go
package hostInfo

import (
	"github.com/docker/engine-api/client"
	"golang.org/x/net/context"
	"math"
	"runtime"
	"strings"
)

type DiskCollector struct {
	unit                float64
	cadvisor            CadvisorAPIClient
	client              *client.Client
	dockerStorageDriver string
}

func (d DiskCollector) convertUnits(number float64) float64 {
	// return in MB
	return math.Floor(number/d.unit*1000) / 1000
}

func (d DiskCollector) getDockerStorageInfo() map[string]interface{} {
	data := map[string]interface{}{}

	if d.client != nil {
		info, _ := d.client.Info(context.Background())
		for _, item := range info.DriverStatus {
			data[item[0]] = item[1]
		}
	}
	return data
}

func (d DiskCollector) includeInFilesystem(device string) bool {
	include := true
	if d.dockerStorageDriver == "devicemapper" {
		pool := d.getDockerStorageInfo()
		poolName, ok := pool["Pool Name"]
		if !ok {
			poolName = "/dev/mapper/docker-"
		}
		if strings.HasSuffix(poolName.(string), "-pool") {
			poolName := poolName.(string)
			poolName = poolName[len(poolName)-5 : len(poolName)]
		}
		if strings.Contains(device, poolName.(string)) {
			include = false
		}
	}
	return include
}

func (d DiskCollector) getMountpointsCadvisor() map[string]interface{} {
	data := map[string]interface{}{}
	stat := d.cadvisor.GetLatestStat()

	if _, ok := stat["filesystem"]; ok {
		for _, fs := range stat["filesystem"].([]map[string]interface{}) {
			device := fs["device"].(string)
			percentUsed := fs["usage"].(float64) / fs["capacity"].(float64) * 100
			data[device] = map[string]interface{}{
				"free":       d.convertUnits(fs["capacity"].(float64) - fs["usage"].(float64)),
				"total":      d.convertUnits(fs["uasge"].(float64)),
				"used":       d.convertUnits(fs["usage"].(float64)),
				"percentage": math.Floor(percentUsed*100) / 100,
			}
		}
	}
	return data
}

func (d DiskCollector) getMachineFilesystemsCadvisor() map[string]interface{} {
	data := map[string]interface{}{}
	machineInfo := d.cadvisor.GetMachineStats()

	if _, ok := machineInfo["filesystems"]; ok {
		for _, filesystem := range machineInfo["filesystems"].([]map[string]interface{}) {
			if d.includeInFilesystem(filesystem["device"].(string)) {
				data[filesystem["device"].(string)] = map[string]interface{}{
					"capacity": d.convertUnits(filesystem["capacity"].(float64)),
				}
			}
		}
	}
	return data
}

func (d DiskCollector) GetData() map[string]interface{} {
	data := map[string]interface{}{
		"fileSystem":                map[string]interface{}{},
		"mountPoints":               map[string]interface{}{},
		"dockerStorageDriverStatus": map[string]interface{}{},
		"dockerStorageDriver":       d.dockerStorageDriver,
	}

	if runtime.GOOS == "linux" {
		for key, value := range d.getMachineFilesystemsCadvisor() {
			data["fileSystems"].(map[string]interface{})[key] = value
		}
		for key, value := range d.getMountpointsCadvisor() {
			data["mountPoints"].(map[string]interface{})[key] = value
		}
	}
	return data
}

func (d DiskCollector) KeyName() string {
	return "diskInfo"
}

func (d DiskCollector) GetLabels(prefix string) map[string]string {
	return map[string]string{}
}

File: tests/tests/cattle/plugins/host_info/disk.py
import platform

from tests.cattle.utils import CadvisorAPIClient
from tests.cattle import Config


class DiskCollector(object):
    def __init__(self, docker_client=None):
        self.unit = 1048576
        self.cadvisor = CadvisorAPIClient(Config.cadvisor_ip(),
                                          Config.cadvisor_port())

        self.docker_client = docker_client
        self.docker_storage_driver = None

        if self.docker_client:
            self.docker_storage_driver = \
                self.docker_client.info().get("Driver", None)

    def _convert_units(self, number):
        # Return in MB
        return round(float(number)/self.unit, 3)

    def _get_dockerstorage_info(self):
        data = {}

        if self.docker_client:
            for item in self.docker_client.info().get("DriverStatus"):
                data[item[0]] = item[1]

        return data

    def _include_in_filesystem(self, device):
        include = True

        if self.docker_storage_driver == "devicemapper":
            pool = self._get_dockerstorage_info()

            pool_name = pool.get("Pool Name", "/dev/mapper/docker-")
            if pool_name.endswith("-pool"):
                pool_name = pool_name[:-5]

            if pool_name in device:
                include = False

        return include

    def _get_mountpoints_cadvisor(self):
        data = {}
        stat = self.cadvisor.get_latest_stat()

        if 'filesystem' in stat.keys():
            for fs in stat['filesystem']:
                device = fs['device']
                percent_used = \
                    float(fs['usage']) / float(fs['capacity']) * 100

                data[device] = {
                    'free': self._convert_units(fs['capacity'] - fs['usage']),
                    'total': self._convert_units(fs['capacity']),
                    'used': self._convert_units(fs['usage']),
                    'percentUsed': round(percent_used, 2)
                }

        return data

    def _get_machine_filesystems_cadvisor(self):
        data = {}
        machine_info = self.cadvisor.get_machine_stats()

        if 'filesystems' in machine_info.keys():
            for filesystem in machine_info['filesystems']:
                if self._include_in_filesystem(filesystem['device']):
                    data[filesystem['device']] = {
                        'capacity': self._convert_units(filesystem['capacity'])
                    }

        return data

    def key_name(self):
        return 'diskInfo'

    def get_data(self):
        data = {
            'fileSystems': {},
            'mountPoints': {},
            'dockerStorageDriverStatus': {},
            'dockerStorageDriver': self.docker_storage_driver
        }

        if platform.system() == 'Linux':
            data['fileSystems'].update(
                self._get_machine_filesystems_cadvisor())
            data['mountPoints'].update(self._get_mountpoints_cadvisor())

        data['dockerStorageDriverStatus'].update(
            self._get_dockerstorage_info())

        return data
